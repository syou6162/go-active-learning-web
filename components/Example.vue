<template>
  <div>
    <b-card no-body>
      <b-card-body
        :title="example | getTitle(75, '...')"
        title-tag="h5"
        class="m-1 p-2"
      >
        <div v-if="example.HatenaBookmark.count > 0">
          <a
            :href="example.HatenaBookmark.entry_url"
            style="color: #ff4166;"
          >{{ example.HatenaBookmark.count }} users</a>:
          <hatena-bookmark-icon 
            v-for="b in example.HatenaBookmark.bookmarks.slice(0, 9)"
            :key="b.user"
            :bookmark="b"
          />
        </div>
        <div v-if="tweets !== undefined && tweets !== null && tweets.length > 0">
          <span style="color: #ff4166;">{{ tweets.length }} mentions</span>:
          <twitter-icon
            v-for="tweet in tweets.slice(0, 8)"
            :key="tweet.ScreenName"
            :tweet="tweet"
          />
        </div>
        Date: {{ example.CreatedAt }}
        <annotate-buttons
          v-if="isAdmin"
          :example="example"
        />
        <b-card-footer>
          <b-button
            :to="example | getExampleUrl"
            class="float-right"
            size="sm"
          >
            Read more
          </b-button>
          <img
            v-if="example.Favicon"
            v-lazy="example.Favicon"
            style="width: 16px; height: 16px;"
            onerror="this.style.display='none'"
          >
          <a :href="example.FinalUrl">{{ example | getDomain }} {{ example | getUserName }}</a>
        </b-card-footer>
      </b-card-body>
    </b-card>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';
import { IsAdmin } from '../plugins/amplify';
import Example from '~/models/Example'
import Tweet from '~/models/Tweet'

@Component({
  components: {
    HatenaBookmarkIcon: () => import('./HatenaBookmarkIcon.vue'),
    TwitterIcon: () => import('./TwitterIcon.vue'),
    AnnotateButtons: () => import('./AnnotateButtons.vue'),
  }
})

export default class ExampleComponent extends Vue {
  modalShow: boolean = false

  @Prop({required: true})
  example!: Object

  @Prop({required: false})
  tweets!: Array<Object>

  @Prop({required: true, default: false})
  isAdmin!: Boolean
}
</script>