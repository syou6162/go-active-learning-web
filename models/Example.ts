import HatenaBookmark from '~/models/HatenaBookmark'
import Bookmark from '~/models/Bookmark'
import ReferringTweets from '~/models/Tweet'

export interface Example {
    Id: number
    Label: number
    Url: string
    FinalUrl: string
    Title: string
    Description: string
    OgDescription: string
    OgType: string
    OgImage: string
    Body: string
    Score: number
    IsNew: boolean
    StatusCode: number
    Favicon: string
    ErrorCount: number
    CreatedAt: string
    UpdatedAt: string
    ReferringTweets: ReferringTweets
    HatenaBookmark: HatenaBookmark
}

function truncate(str: string, length: number, omission: string): string {
    str = str ? str : '';
    if (str.length <= length) {
        return str;
    }
    else {
        return str.substring(0, length) + omission;
    }
}

export function getTitle(example: Example, length: number, omission: string): string {
    var title = example.Title ? example.Title : example.Url;
    return truncate(title, length, omission);
}

export function getDomain(example: Example): string {
    var url = example.FinalUrl;
    return url.replace('http://', '').replace('https://', '').split(/[/?#]/)[0];
}

export function getUserName(example: Example): string {
    var domain = getDomain(example);
    var url = example.FinalUrl;
    var paths = url.replace('http://', '').replace('https://', '').split(/[/?#]/);
    if (paths.length === 0) {
        return "";
    } else if ('twitter.com' === domain) {
        return '(@' + paths[1] + ')';
    } else if ('github.com' === domain) {
        return '(@' + paths[1] + ')';
    } else if ('qiita.com' === domain) {
        return '(@' + paths[1] + ')';
    } else if ('www.slideshare.net' === domain) {
        return '(id:' + paths[1] + ')';
    } else if ('speakerdeck.com' === domain) {
        return '(id:' + paths[1] + ')';
    } else {
        return "";
    }
}

export function getDescription(example: Example, length: number, omission: string) {
    var title = example.Title ? example.Title : example.Url;
    var body = example.Body ? example.Body : title;
    var desc = example.OgDescription ? example.OgDescription : (example.Description ? example.Description : body);
    return truncate(desc, length, omission);
}

export function filterBookmarksWithComment(example: Example): Bookmark[] {
    return example.HatenaBookmark.bookmarks.filter(function (b: Bookmark) {
        return b.comment !== "";
    });
}

export function getDescriptionForSearchEngine(example: Example) {
    var tweets = example.ReferringTweets.Tweets.map(t => "@" + t.ScreenName + "「" + t.FullText.substr(0, 100) + "...」").slice(0, 3);
    var bookmarks = filterBookmarksWithComment(example).map(b => "id:" + b.user + "「" + b.comment + "」").slice(0, 3);
    return tweets.join("\n") + bookmarks.join("\n");
}

export function getExampleUrl(example: Example) {
    return '/example/' + example.Id;
}