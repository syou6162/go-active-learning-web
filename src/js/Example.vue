<template>
  <div>
    <b-card no-body>
      <b-card-body v-bind:title="example | getTitle(75, '...')" title-tag="h5" class="m-1 p-2">
        <div v-if="example.HatenaBookmark.count > 0">
          <a v-bind:href="example.HatenaBookmark.entry_url" style="color: #ff4166;">{{ example.HatenaBookmark.count }} users</a>:
          <hatena-bookmark-icon 
            v-for="b in example.HatenaBookmark.bookmarks.slice(0, 9)"
            v-bind:bookmark="b"
            v-bind:key="b.user"
            ></hatena-bookmark-icon>
        </div>
        <div v-if="tweets !== undefined && tweets !== null && tweets.length > 0">
          <span style="color: #ff4166;">{{ tweets.length }} mentions</span>:
          <twitter-icon
             v-for="tweet in tweets.slice(0, 8)"
             v-bind:tweet="tweet"
             v-bind:key="tweet.ScreenName"
             ></twitter-icon>
        </div>
        Date: 
        {{ example.CreatedAt.tz("Asia/Tokyo").format("YYYY/MM/DD HH:mm") }}
        <annotate-buttons
          v-if="isAdmin"
          v-bind:example="example"
          ></annotate-buttons>
        <b-card-footer>
          <b-button v-bind:href="'/example/' + encodeURIComponent(example.FinalUrl)" class="float-right" size="sm" v-bind:variant="example | getButtonStyle">Read more</b-button>
          <img v-if="example.Favicon" style="width: 16px; height: 16px;" v-lazy="example.Favicon" onerror="this.style.display='none'" />
          <a v-bind:href="example.FinalUrl">{{ example | getDomain }} {{ example | getUserName }}</a>
        </b-card-footer>
      </b-card-body>
    </b-card>
  </div>
</template>

<script>
import HatenaBookmarkIcon from './HatenaBookmarkIcon.vue';
import TwitterIcon from './TwitterIcon.vue';
import AnnotateButtons from './AnnotateButtons.vue';

export default {
  data () {
    return {
      modalShow: false
    }
  },
  props: ['example', 'tweets', 'isAdmin'],
  components: {
    "hatena-bookmark-icon": HatenaBookmarkIcon,
    "twitter-icon": TwitterIcon,
    "annotate-buttons": AnnotateButtons 
  }
}
</script>
