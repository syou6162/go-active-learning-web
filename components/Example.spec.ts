import { createLocalVue, mount } from '@vue/test-utils'
import { BButton, BCard, BCardBody, BCardFooter } from 'bootstrap-vue'
import Example from './Example.vue'

const localVue = createLocalVue();
localVue.component('b-button', BButton)
localVue.component('b-card', BCard)
localVue.component('b-card-body', BCardBody)
localVue.component('b-card-footer', BCardFooter)

const hatenaBookmark = {
  title: "Embulk & Digdag Online Meetup 2020【イベントレポート】 | trocco(トロッコ)",
  screenshot: "https://b.st-hatena.com/images/v4/public/common/noimage.png?version=8358883de4e4eced6d5c2b49020f40bbbcfdca71",
  entry_url: "https://b.hatena.ne.jp/entry/s/blog.trocco.io/dev-articles/embulk-digdag-meetup-2020",
  bookmarks: [
    {
      "comment": "",
      "user": "machupicchubeta",
      "tags": [],
      "timestamp": "2020/08/27 00:56"
    },
    {
      "tags": [],
      "timestamp": "2020/08/26 19:53",
      "user": "hoppie",
      "comment": ""
    },
    {
      "comment": "“統合”",
      "user": "polamjag",
      "timestamp": "2020/08/26 14:29",
      "tags": []
    },
    {
      "comment": "",
      "user": "yukiyan_w",
      "tags": [
        "digdag",
        "embulk"
      ],
      "timestamp": "2020/08/26 12:15"
    }
  ],
  count: 5,
  url: "https://blog.trocco.io/dev-articles/embulk-digdag-meetup-2020",
  eid: "4690501516113184834"
}

describe('Example', () => {
  test('ラベルが0のときはunlabeldが有効になっている', () => {
    const wrapper = mount(Example, {
      localVue,
      propsData: {
        example: {
          Id: 11111,
          Label: 0,
          Url: "https://blog.trocco.io/dev-articles/embulk-digdag-meetup-2020",
          FinalUrl: "https://blog.trocco.io/dev-articles/embulk-digdag-meetup-2020",
          Title: "Embulk & Digdag Online Meetup 2020【イベントレポート】 | trocco(トロッコ)",
          Description: "",
          HatenaBookmark: hatenaBookmark,
        },
        tweets: {
          Tweets: [],
          Count: 0
        },
        isAdmin: false
      }
    })
    expect(wrapper.exists()).toBeTruthy()
    expect(wrapper.find('h5').classes()).toContain("card-title")
    expect(wrapper.element).toMatchSnapshot()
  })
})