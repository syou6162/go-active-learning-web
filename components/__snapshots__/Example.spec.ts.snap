// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example ラベルが0のときはunlabeldが有効になっている 1`] = `
<div>
  <div
    class="card"
  >
    <!---->
    <!---->
    <div
      class="card-body m-1 p-2"
    >
      <h5
        class="card-title"
      >
        Embulk & Digdag Online Meetup 2020【イベントレポート】 | trocco(トロッコ)
      </h5>
      <!---->
      <p
        class="example-description"
      >
        
        Embulk & Digdag Online Meetup 2020【イベントレポート】 | trocco(トロッコ)
      
      </p>
       
      <a
        class="hatena-bookmark-link"
        href="https://b.hatena.ne.jp/entry/s/blog.trocco.io/dev-articles/embulk-digdag-meetup-2020"
        rel="noopener"
        target="_blank"
      >
        5 users
      </a>
      ,
      
      <span
        class="tweets-count"
      >
        0 mentions
      </span>
       
      <span
        class="example-created-at"
      >
        
      </span>
       
      <!---->
       
      <div
        class="card-footer"
      >
        <a
          class="btn float-right btn-secondary btn-sm"
          href="/example/11111"
          target="_self"
        >
          
          Read more
        
        </a>
         
        <img
          alt="Embulk & Digdag Online Meetup 2020【イベントレポート】 | trocco(トロッコ)"
          class="example-favicon-img"
          src="/img/website_icon.png"
        />
         
        <a
          href="https://blog.trocco.io/dev-articles/embulk-digdag-meetup-2020"
          rel="noopener"
          target="_blank"
        >
          blog.trocco.io 
        </a>
      </div>
    </div>
    <!---->
    <!---->
  </div>
</div>
`;
